Excessive scrolling gives runtime error for DateSeparator with the jump_to_date feature enabled

### Steps to reproduce

Scroll up/down to render DateSeparators with the jump_to_date feature enabled. Sooner or later the follwing runtime error will be displayed in the console and the DateSeparator will stop working.

rageshake.ts:69 Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at getRootForUpdatedFiber (react-dom-client.development.js:4624:1)
    at enqueueConcurrentHookUpdate (react-dom-client.development.js:4584:1)
    at dispatchSetStateInternal (react-dom-client.development.js:9167:1)
    at dispatchSetState (react-dom-client.development.js:9127:1)
    at floating-ui.react.mjs:2949:1
    at ref.current (floating-ui.react.mjs:26:1)
    at floating-ui.react.mjs:35:1
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (floating-ui.react.mjs:35:1)
    at floating-ui.react.mjs:45:1

The above error occurred in the <div> component.

React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.

consoleObj.<computed> @ rageshake.ts:69
defaultOnCaughtError @ react-dom-client.development.js:9410
logCaughtError @ react-dom-client.development.js:9446
runWithFiberInDEV @ react-dom-client.development.js:871
inst.componentDidCatch.update.callback @ react-dom-client.development.js:9493
callCallback @ react-dom-client.development.js:7423
commitCallbacks @ react-dom-client.development.js:7443
runWithFiberInDEV @ react-dom-client.development.js:871
commitClassCallbacks @ react-dom-client.development.js:13377
commitLayoutEffectOnFiber @ react-dom-client.development.js:14026
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14031
flushLayoutEffects @ react-dom-client.development.js:18138
commitRoot @ react-dom-client.development.js:17954
commitRootWhenReady @ react-dom-client.development.js:16824
performWorkOnRoot @ react-dom-client.development.js:16722
performSyncWorkOnRoot @ react-dom-client.development.js:18972
flushSyncWorkAcrossRoots_impl @ react-dom-client.development.js:18814
processRootScheduleInMicrotask @ react-dom-client.development.js:18853
(anonymous) @ react-dom-client.development.js:18991Understand this error
rageshake.ts:69 Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at getRootForUpdatedFiber (react-dom-client.development.js:4624:1)
    at enqueueConcurrentHookUpdate (react-dom-client.development.js:4584:1)
    at dispatchSetStateInternal (react-dom-client.development.js:9167:1)
    at dispatchSetState (react-dom-client.development.js:9127:1)
    at floating-ui.react.mjs:2949:1
    at ref.current (floating-ui.react.mjs:26:1)
    at floating-ui.react.mjs:35:1
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (floating-ui.react.mjs:35:1)
    at floating-ui.react.mjs:45:1
consoleObj.<computed> @ rageshake.ts:69
(anonymous) @ logger.ts:114
componentDidCatch @ ErrorBoundary.tsx:47
react_stack_bottom_frame @ react-dom-client.development.js:25959
inst.componentDidCatch.update.callback @ react-dom-client.development.js:9504
callCallback @ react-dom-client.development.js:7423
commitCallbacks @ react-dom-client.development.js:7443
runWithFiberInDEV @ react-dom-client.development.js:871
commitClassCallbacks @ react-dom-client.development.js:13377
commitLayoutEffectOnFiber @ react-dom-client.development.js:14026
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14065
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13949
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:13954
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14165
recursivelyTraverseLayoutEffects @ react-dom-client.development.js:15159
commitLayoutEffectOnFiber @ react-dom-client.development.js:14031
flushLayoutEffects @ react-dom-client.development.js:18138
commitRoot @ react-dom-client.development.js:17954
commitRootWhenReady @ react-dom-client.development.js:16824
performWorkOnRoot @ react-dom-client.development.js:16722
performSyncWorkOnRoot @ react-dom-client.development.js:18972
flushSyncWorkAcrossRoots_impl @ react-dom-client.development.js:18814
processRootScheduleInMicrotask @ react-dom-client.development.js:18853
(anonymous) @ react-dom-client.development.js:18991Understand this error
rageshake.ts:69 The above error occurred while React was rendering the following components: 
    at div (<anonymous>)
    at J (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:33379:18)
    at TooltipAnchor (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:20190:3)
    at Tooltip (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:20123:7)
    at Wx (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:39785:22)
    at Primitive.button.SlotClone (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:17542:13)
    at Primitive.button.Slot (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:17518:13)
    at Primitive.button (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:16969:13)
    at Primitive.div.SlotClone (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:17542:13)
    at Primitive.div.Slot (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:17518:13)
    at Primitive.div (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:16969:13)
    at PopperAnchor (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:16466:13)
    at MenuAnchor (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:15660:13)
    at DropdownMenuTrigger (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:14450:13)
    at Provider (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:13639:15)
    at Provider (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:13639:15)
    at Provider (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:13639:15)
    at Popper (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:16458:11)
    at Menu (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:15614:11)
    at Provider (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:13639:15)
    at DropdownMenu (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:14415:5)
    at Menu (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:19616:3)
    at Ux (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:39740:16)
    at div (<anonymous>)
    at J (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:33379:18)
    at jx (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:39670:116)
    at Gx (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:39808:19)
    at div (<anonymous>)
    at div (<anonymous>)
    at http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:23193:368
    at div (<anonymous>)
    at mr (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:23377:25)
    at div (<anonymous>)
    at http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:23326:71
    at http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:23398:13
    at http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:23046:23
    at sS (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:40010:19)
    at div (<anonymous>)
    at TimelinePanelViewContent (http://localhost:8080/bundles/_dev_/element-web-app.js:48694:3)
    at TimelinePanelView (http://localhost:8080/bundles/_dev_/element-web-app.js:48680:3)
    at TimelinePanel (http://localhost:8080/bundles/_dev_/element-web-app.js:17256:5)
    at main (<anonymous>)
    at div (<anonymous>)
    at div (<anonymous>)
    at MainSplit (http://localhost:8080/bundles/_dev_/element-web-app.js:8075:5)
    at ErrorBoundary (http://localhost:8080/bundles/_dev_/init.js:35482:5)
    at div (<anonymous>)
    at http://localhost:8080/bundles/_dev_/init.js:57607:7
    at RoomView (http://localhost:8080/bundles/_dev_/element-web-app.js:12965:5)
    at div (<anonymous>)
    at div (<anonymous>)
    at div (<anonymous>)
    at qt (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:26145:13)
    at div (<anonymous>)
    at Ut (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:25845:13)
    at MC (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:41727:19)
    at div (<anonymous>)
    at div (<anonymous>)
    at MatrixClientContextProvider (http://localhost:8080/bundles/_dev_/element-web-app.js:10365:61)
    at LoggedInView (http://localhost:8080/bundles/_dev_/element-web-app.js:7429:5)
    at FloatingDelayGroup (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:7989:5)
    at TooltipProvider (http://localhost:8080/bundles/_dev_/default-src_audio_consts_ts.js:20309:3)
    at ErrorBoundary (http://localhost:8080/bundles/_dev_/init.js:35482:5)
    at MatrixChat (http://localhost:8080/bundles/_dev_/element-web-app.js:8389:5)

### Outcome

#### What did you expect?
To be able to srcoll up/down without problems
#### What happened instead?
A runtime error was displayed in the console and the DateSeparators stopped working

### Operating system

_No response_

### Browser information

_No response_

### URL for webapp

_No response_

### Application version

_No response_

### Homeserver

_No response_

### Will you send logs?

No

--- a/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
+++ b/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
@@ -6,7 +6,6 @@
  */
 
 import React, { forwardRef } from "react";
-import { mergeRefs } from "react-merge-refs";
 import { Heading, Tooltip } from "@vector-im/compound-web";
 import ChevronDownIcon from "@vector-im/compound-design-tokens/assets/web/icons/chevron-down";
 
@@ -21,8 +20,6 @@ export interface DateSeparatorButtonProps {
     tooltipOpen?: boolean;
     /** Extra CSS classes to apply to the component. */
     className?: string;
-    /** Optional ref for the button container element. */
-    buttonRef?: React.Ref<HTMLDivElement>;
     /** Called when the pointer enters the button trigger. */
     onMouseEnter?: React.MouseEventHandler<HTMLDivElement>;
     /** Called when the pointer leaves the button trigger. */
@@ -35,7 +32,7 @@ export interface DateSeparatorButtonProps {
 
 /** Interactive date separator button that opens the jump-to-date menu. */
 export const DateSeparatorButton = forwardRef<HTMLDivElement, DateSeparatorButtonProps>(function DateSeparatorButton(
-    { label, tooltipOpen, className, buttonRef, ...props },
+    { label, tooltipOpen, className, ...props },
     forwardedRef,
 ): React.ReactNode {
     const { translate: _t } = useI18n();
@@ -43,7 +40,7 @@ export const DateSeparatorButton = forwardRef<HTMLDivElement, DateSeparatorButto
     return (
         <Tooltip description={_t("room|jump_to_date")} placement="right" open={tooltipOpen}>
             <Flex
-                ref={mergeRefs([buttonRef, forwardedRef])}
+                ref={forwardedRef}
                 data-testid="jump-to-date-separator-button"
                 className={className}
                 aria-live="off"

--- a/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
+++ b/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
@@ -5,7 +5,8 @@
  * Please see LICENSE files in the repository root for full details.
  */
 
-import React from "react";
+import React, { forwardRef } from "react";
+import { mergeRefs } from "react-merge-refs";
 import { Heading, Tooltip } from "@vector-im/compound-web";
 import ChevronDownIcon from "@vector-im/compound-design-tokens/assets/web/icons/chevron-down";
 
@@ -33,18 +34,16 @@ export interface DateSeparatorButtonProps {
 }
 
 /** Interactive date separator button that opens the jump-to-date menu. */
-export function DateSeparatorButton({
-    label,
-    tooltipOpen,
-    className,
-    buttonRef,
-    ...props
-}: DateSeparatorButtonProps): React.ReactNode {
+export const DateSeparatorButton = forwardRef<HTMLDivElement, DateSeparatorButtonProps>(function DateSeparatorButton(
+    { label, tooltipOpen, className, buttonRef, ...props },
+    forwardedRef,
+): React.ReactNode {
     const { translate: _t } = useI18n();
+
     return (
         <Tooltip description={_t("room|jump_to_date")} placement="right" open={tooltipOpen}>
             <Flex
-                ref={buttonRef}
+                ref={mergeRefs([buttonRef, forwardedRef])}
                 data-testid="jump-to-date-separator-button"
                 className={className}
                 aria-live="off"
@@ -60,4 +59,4 @@ export function DateSeparatorButton({
             </Flex>
         </Tooltip>
     );
-}
+});

--- a/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
+++ b/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorButton.tsx
@@ -30,7 +30,7 @@ export interface DateSeparatorButtonProps {
     onBlur?: React.FocusEventHandler<HTMLDivElement>;
 }
 
-/** Interactive date separator button that opens the jump-to-date menu. */
+/** Interactive date separator button that forwards its ref to the underlying trigger element used by Compound Menu. */
 export const DateSeparatorButton = forwardRef<HTMLDivElement, DateSeparatorButtonProps>(function DateSeparatorButton(
     { label, tooltipOpen, className, ...props },
     forwardedRef,

--- a/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorView.test.tsx
+++ b/packages/shared-components/src/room/timeline/DateSeparatorView/DateSeparatorView.test.tsx
@@ -7,12 +7,13 @@
 
 import { render, screen } from "@test-utils";
 import { composeStories } from "@storybook/react-vite";
-import React from "react";
+import React, { createRef } from "react";
 import { describe, it, expect } from "vitest";
 import { waitFor } from "@testing-library/react";
 import userEvent from "@testing-library/user-event";
 
 import { BaseViewModel } from "../../../core/viewmodel/BaseViewModel";
+import { DateSeparatorButton } from "./DateSeparatorButton";
 import { DateSeparatorView, type DateSeparatorViewModel, type DateSeparatorViewSnapshot } from "./DateSeparatorView";
 import * as stories from "./DateSeparatorView.stories";
 
@@ -72,5 +73,13 @@ describe("DateSeparatorView", () => {
         expect(screen.getByText("Today", { selector: "h2" })).toBeInTheDocument();
         vm.setSnapshot({ label: "Yesterday" });
         await waitFor(() => expect(screen.getByText("Yesterday", { selector: "h2" })).toBeInTheDocument());
+    });
+
+    it("forwards refs to the trigger element", () => {
+        const ref = createRef<HTMLDivElement>();
+
+        render(<DateSeparatorButton ref={ref} label="Today" />);
+
+        expect(ref.current).toBe(screen.getByTestId("jump-to-date-separator-button"));
     });
 });

